ports/zephyr/zephyr_display.c

```diff
@@ -0,0 +1,353 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/display.h>
+#include <stdlib.h>
+
+#include "modzephyr.h"
+#include "zephyr_device.h"
+#include "py/runtime.h"
+
+#if defined(CONFIG_DISPLAY)
+
+#if DT_HAS_COMPAT_STATUS_OKAY(zephyr_displays)
+#define ENUMERATE_DISPLAY_DEVS(node_id, prop, idx) DEVICE_DT_GET(DT_PROP_BY_IDX(node_id, prop, idx)),
+
+static const struct device *zephyr_display_devices[DT_ZEPHYR_DISPLAYS_COUNT] = {
+    DT_FOREACH_PROP_ELEM(DT_COMPAT_GET_ANY_STATUS_OKAY(zephyr_displays), displays, ENUMERATE_DISPLAY_DEVS)
+};
+#elif DT_HAS_CHOSEN(zephyr_display)
+static const struct device *zephyr_display_devices[] = {
+    DEVICE_DT_GET(DT_CHOSEN(zephyr_display)),
+};
+#else
+static const struct device *zephyr_display_devices[1];
+#endif
+
+const mp_obj_type_t zephyr_display_type;
+
+typedef struct _zephyr_display_obj_t {
+    mp_obj_base_t base;
+    const struct device *dev;
+} zephyr_display_obj_t;
+
+static void zephyr_display_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    zephyr_display_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "Display(\"%s\")", self->dev->name);
+}
+
+static mp_obj_t zephyr_display_make_new(const mp_obj_type_t *type,
+    size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+    const struct device *dev;
+
+    if (mp_obj_is_str(args[0])) {
+        dev = zephyr_device_find(args[0]);
+    } else {
+        int id = mp_obj_get_int(args[0]);
+
+        if (id < 0 || id >= DT_ZEPHYR_DISPLAYS_COUNT) {
+            mp_raise_ValueError(MP_ERROR_TEXT("Invalid display ID"));
+            return mp_const_none;
+        }
+        dev = zephyr_display_devices[id];
+    }
+    if (!device_is_ready(dev)) {
+        mp_raise_msg(&mp_type_RuntimeError, MP_ERROR_TEXT("Display is not ready"));
+        return mp_const_none;
+    }
+
+    zephyr_display_obj_t *self = mp_obj_malloc(zephyr_display_obj_t, type);
+    self->dev = dev;
+    return MP_OBJ_FROM_PTR(self);
+}
+
+static mp_obj_t zephyr_display_write(size_t n_args, const mp_obj_t *args_in) {
+    zephyr_display_obj_t *self = MP_OBJ_TO_PTR(args_in[0]);
+    struct display_capabilities capabilities;
+    int ret;
+    size_t len;
+    mp_buffer_info_t bufinfo;
+    int x = mp_obj_get_int(args_in[2]);
+    int y = mp_obj_get_int(args_in[3]);
+    int sx;
+    int sy;
+
+    mp_get_buffer_raise(args_in[1], &bufinfo, MP_BUFFER_READ);
```

Comment:
At [line 134](https://github.com/micropython/micropython/pull/17939/files#diff-52d15353d62a3e3d31e86d070c43f96560bb080eb9fc5080a42ed1d85a4a21f7R134) it is possible to have between 2 and 6 arguments.

Maybe you meant to allow `disp.write(fb)` and `disp.write(fb, 0)`?

```suggestion
    int x = 0;
    int y = 0;
    int sx;
    int sy;

    mp_get_buffer_raise(args_in[1], &bufinfo, MP_BUFFER_READ);

    if (n_args > 2) {
        x = mp_obj_get_int(args_in[2]);
    }

    if (n_args > 3) {
        y = mp_obj_get_int(args_in[3]);
    }
```

Reply:
Yes, that is a mistake, it should be able to take only the buffer argument.